// @grafana/faro-core/dist/cjs/api/meta/initialize.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.initializeMetaAPI = void 0;
function initializeMetaAPI(_unpatchedConsole, internalLogger, _config, metas, _transports) {
    internalLogger.debug('Initializing meta API');
    var metaSession = undefined;
    var metaUser = undefined;
    var metaView = undefined;
    var setUser = function (user) {
        if (metaUser) {
            metas.remove(metaUser);
        }
        metaUser = {
            user: user,
        };
        metas.add(metaUser);
    };
    var setSession = function (session) {
        if (metaSession) {
            metas.remove(metaSession);
        }
        metaSession = {
            session: session,
        };
        metas.add(metaSession);
    };
    var getSession = function () { return metas.value.session; };
    var setView = function (view) {
        var _a;
        if (((_a = metaView === null || metaView === void 0 ? void 0 : metaView.view) === null || _a === void 0 ? void 0 : _a.name) === (view === null || view === void 0 ? void 0 : view.name)) {
            return;
        }
        var previousView = metaView;
        metaView = {
            view: view,
        };
        metas.add(metaView);
        if (previousView) {
            metas.remove(previousView);
        }
    };
    var getView = function () { return metas.value.view; };
    return {
        setUser: setUser,
        resetUser: setUser,
        setSession: setSession,
        resetSession: setSession,
        getSession: getSession,
        setView: setView,
        getView: getView,
    };
}
exports.initializeMetaAPI = initializeMetaAPI;
