// @grafana/schema/dist/esm/veneer/common.types.js
import { TableCellDisplayMode } from '../common/common.gen.js';
export { AxisColorMode, AxisPlacement, BarAlignment, BarGaugeDisplayMode, BarGaugeValueMode, BigValueColorMode, BigValueGraphMode, BigValueJustifyMode, BigValueTextMode, ComparisonOperation, FrameGeometrySourceMode, GraphDrawStyle, GraphGradientMode, GraphTransform, GraphTresholdsStyleMode, HeatmapCalculationMode, HeatmapCellLayout, LegendDisplayMode, LineInterpolation, LogsDedupStrategy, LogsSortOrder, ScaleDirection, ScaleDistribution, ScaleOrientation, SortOrder, StackingMode, TableCellBackgroundDisplayMode, TableCellDisplayMode, TableCellHeight, TextDimensionMode, TooltipDisplayMode, VariableFormatID, VisibilityMode, VizOrientation, defaultLineStyle, defaultOptionsWithTimezones, defaultReduceDataOptions, defaultTableFooterOptions, defaultTimeZone, defaultVizLegendOptions } from '../common/common.gen.js';

const defaultTableFieldOptions = {
  align: "auto",
  inspect: false,
  cellOptions: {
    type: TableCellDisplayMode.Auto
  }
};

export { defaultTableFieldOptions };
