// @grafana/ui/dist/esm/components/Forms/Legacy/Switch/Switch.js
import { uniqueId } from 'lodash';
import React__default, { PureComponent } from 'react';
import '../../../FormField/FormField.js';
import '../Input/Input.js';
import 'react-select';
import { Icon } from '../../../Icon/Icon.js';
import '../Select/Select.js';
import { Tooltip } from '../../../Tooltip/Tooltip.js';
import '../../../SecretFormField/SecretFormField.js';
import '../../../IconButton/IconButton.js';
import '../../../ConfirmButton/ConfirmButton.js';
import '../../../Button/Button.js';
import '../../../Button/ButtonGroup.js';
import 'react-popper';
import 'react-transition-group/Transition';
import '../../../Portal/Portal.js';
import '../../../Toggletip/Toggletip.js';
import '@emotion/css';
import 'react-custom-scrollbars-2';
import '../../../../themes/ThemeContext.js';
import '@grafana/data';
import 'memoize-one';
import '@emotion/react';
import 'tinycolor2';
import 'classnames';
import '../../../TabbedContainer/TabbedContainer.js';
import '../../../../utils/i18n.js';
import '../../../Cascader/Cascader.js';
import '../../../ButtonCascader/ButtonCascader.js';
import '../../../Spinner/Spinner.js';
import '../../../ColorPicker/ColorPicker.js';
import '../../../ColorPicker/ColorPickerInput.js';
import '../../../ColorPicker/SeriesColorPickerPopover.js';
import '../../../StatsPicker/StatsPicker.js';
import '../../../RefreshPicker/RefreshPicker.js';
import '../../../DateTimePickers/TimeRangePicker.js';
import 'rc-time-picker';
import 'react-select/async';
import 'react-select/async-creatable';
import 'react-select/creatable';
import '../../../Select/IndicatorsContainer.js';
import '../../../Select/InputControl.js';
import '../../../Select/getSelectStyles.js';
import '../../../Select/SelectContainer.js';
import '../../../Select/SelectMenu.js';
import '../../../Select/SelectOptionGroup.js';
import '../../../transitions/FadeTransition.js';
import '../../../transitions/SlideOutTransition.js';
import '../../../Select/ValueContainer.js';
import '../../../DateTimePickers/TimeZonePicker/TimeZoneOption.js';
import '@grafana/e2e-selectors';
import '../../../DateTimePickers/DatePicker/DatePicker.js';
import '../../../Input/Input.js';
import '../../../DateTimePickers/DateTimePicker/DateTimePicker.js';
import '../../../List/AbstractList.js';
import 'react-table';
import '../../../InteractiveTable/ExpanderCell.js';
import '../../../../utils/dom.js';
import '../../../../utils/colors.js';
import 'slate';
import 'ansicolor';
import '../../../../utils/logger.js';
import '../../../Tags/Tag.js';
import '../../../Tags/TagList.js';
import '../../../Modal/Modal.js';
import '../../../QueryField/QueryField.js';
import '../../../Monaco/CodeEditor.js';
import 'react-use';
import '../../../Modal/getModalStyles.js';
import '../../../Tabs/Tab.js';
import '../../../Tabs/TabsBar.js';
import '../../../Modal/ModalsContext.js';
import '../../../PageLayout/PageToolbar.js';
import 'rxjs';
import 'rxjs/operators';
import '../../../Table/Table.js';
import '@grafana/schema';
import '../../../TableInputCSV/TableInputCSV.js';
import '../../../Tabs/VerticalTab.js';
import '../../../Tabs/TabContent.js';
import '../../../Tabs/Counter.js';
import '../../../BigValue/BigValue.js';
import '../../../Sparkline/Sparkline.js';
import '../../../Gauge/Gauge.js';
import '../../../Graph/Graph.js';
import '../../../Graph/GraphWithLegend.js';
import '../../../ContextMenu/ContextMenu.js';
import '../../../FormattedValueDisplay/FormattedValueDisplay.js';
import '../../../Menu/MenuGroup.js';
import '../../../Menu/MenuItem.js';
import '../../../VizLegend/SeriesIcon.js';
import '../../../BarGauge/BarGauge.js';
import '../../../VizTooltip/VizTooltip.js';
import '../../../VizTooltip/VizTooltipContainer.js';
import '../../../VizRepeater/VizRepeater.js';
import '../../../PanelChrome/index.js';
import '../../../VizLayout/VizLayout.js';
import '../../../VizLegend/VizLegend.js';
import '../../../VizLegend/VizLegendListItem.js';
import '../../../Alert/Alert.js';
import '../../../Collapse/Collapse.js';
import '../../../Logs/LogMessageAnsi.js';
import '../../../Logs/LogRows.js';
import '../../../Button/FullWidthButtonContainer.js';
import '../../../ClickOutsideWrapper/ClickOutsideWrapper.js';
import '../../../Menu/Menu.js';
import '../../../DataLinks/DataLinksInlineEditor/DataLinksInlineEditor.js';
import '../../../DataLinks/DataLinkInput.js';
import '../../../InfoBox/InfoBox.js';
import '../../../InfoBox/FeatureInfoBox.js';
import '../../../JSONFormatter/JSONFormatter.js';
import '../../../ErrorBoundary/ErrorBoundary.js';
import '../../../ErrorBoundary/ErrorWithStack.js';
import '../../InlineField.js';
import '../../../Switch/Switch.js';
import '../../../DataSourceSettings/CustomHeadersSettings.js';
import '../../../TextArea/TextArea.js';
import 'react-use/lib/useClickAway';
import '@react-aria/dialog';
import '@react-aria/focus';
import '@react-aria/overlays';
import 'rc-drawer';
import '../../../Slider/Slider.js';
import '../../../Slider/RangeSlider.js';
import 'react-hook-form';
import '../../../ToolbarButton/ToolbarButton.js';
import '../../../ToolbarButton/ToolbarButtonRow.js';
import '../../../MatchersUI/fieldMatchersUI.js';
import '../../../Link/Link.js';
import '../../Label.js';
import '../../Field.js';
import '../../FieldSet.js';
import '../../FieldValidationMessage.js';
import '../../InlineSegmentGroup.js';
import '../../../Badge/Badge.js';
import '../../RadioButtonGroup/RadioButtonGroup.js';
import '../../../Input/AutoSizeInput.js';
import '../../../FilterInput/FilterInput.js';
import '../../Checkbox.js';
import 'uuid';
import 'react-dropzone';
import '../../../DateTimePickers/TimeRangeInput.js';
import '../../../DateTimePickers/RelativeTimeRangePicker/RelativeTimeRangePicker.js';
import '../../../Card/Card.js';
import '../../../Card/CardContainer.js';
import '../../../Dropdown/ButtonSelect.js';
import '../../../Dropdown/Dropdown.js';
import '../../../PluginSignatureBadge/PluginSignatureBadge.js';
import '../../../uPlot/config.js';
import 'uplot';
import '../../../uPlot/utils.js';
import '../../../uPlot/PlotLegend.js';
import '../../../GraphNG/GraphNG.js';
import '../../../TimeSeries/TimeSeries.js';
import '../../../GraphNG/hooks.js';
import '../../../BrowserLabel/Label.js';

class Switch extends PureComponent {
  constructor() {
    super(...arguments);
    this.state = {
      id: uniqueId()
    };
    this.internalOnChange = (event) => {
      event.stopPropagation();
      this.props.onChange(event);
    };
  }
  render() {
    const {
      labelClass = "",
      switchClass = "",
      label,
      checked,
      disabled,
      transparent,
      className,
      tooltip,
      tooltipPlacement
    } = this.props;
    const labelId = this.state.id;
    const labelClassName = `gf-form-label ${labelClass} ${transparent ? "gf-form-label--transparent" : ""} pointer`;
    const switchClassName = `gf-form-switch ${switchClass} ${transparent ? "gf-form-switch--transparent" : ""}`;
    return /* @__PURE__ */ React__default.createElement("div", { className: "gf-form-switch-container-react" }, /* @__PURE__ */ React__default.createElement("label", { htmlFor: labelId, className: `gf-form gf-form-switch-container ${className || ""}` }, label && /* @__PURE__ */ React__default.createElement("div", { className: labelClassName }, label, tooltip && /* @__PURE__ */ React__default.createElement(Tooltip, { placement: tooltipPlacement ? tooltipPlacement : "auto", content: tooltip, theme: "info" }, /* @__PURE__ */ React__default.createElement("div", { className: "gf-form-help-icon gf-form-help-icon--right-normal" }, /* @__PURE__ */ React__default.createElement(Icon, { name: "info-circle", size: "sm", style: { marginLeft: "10px" } })))), /* @__PURE__ */ React__default.createElement("div", { className: switchClassName }, /* @__PURE__ */ React__default.createElement(
      "input",
      {
        disabled,
        id: labelId,
        type: "checkbox",
        checked,
        onChange: this.internalOnChange
      }
    ), /* @__PURE__ */ React__default.createElement("span", { className: "gf-form-switch__slider" }))));
  }
}

export { Switch };
