// @grafana/faro-core/dist/esm/index.js
export { defaultExceptionType } from './api';
export { BaseExtension } from './extensions';
export { globalObject } from './globalObject';
export { initializeFaro } from './initialize';
export { defaultBatchingConfig, defaultGlobalObjectKey } from './config';
export { BaseInstrumentation } from './instrumentations';
export { createInternalLogger, defaultInternalLoggerLevel, InternalLoggerLevel } from './internalLogger';
export { faro, getInternalFaroFromGlobalObject, internalGlobalObjectKey, isInternalFaroOnGlobalObject, setInternalFaroOnGlobalObject, } from './sdk';
export * from './semantic';
export { BaseTransport, getTransportBody, TransportItemType, transportItemTypeToBodyKey } from './transports';
export { defaultUnpatchedConsole } from './unpatchedConsole';
export { allLogLevels, createPromiseBuffer, deepEqual, defaultLogLevel, genShortID, getCurrentTimestamp, isArray, isBoolean, isDomError, isDomException, isElement, isElementDefined, isError, isErrorDefined, isErrorEvent, isEvent, isEventDefined, isFunction, isInstanceOf, isInt, isMap, isMapDefined, isNull, isNumber, isObject, isPrimitive, isRegExp, isString, isSymbol, isSyntheticEvent, isThenable, isToString, isTypeof, isUndefined, LogLevel, noop, dateNow, } from './utils';
export { VERSION } from './version';
