// @grafana/ui/dist/esm/components/Logs/LogLabelStatsRow.js
import { cx, css } from '@emotion/css';
import React__default from 'react';
import { useStyles2 } from '../../themes/ThemeContext.js';

const getStyles = (theme) => ({
  logsStatsRow: css`
    label: logs-stats-row;
    margin: ${parseInt(theme.spacing(2), 10) / 1.75}px 0;
  `,
  logsStatsRowActive: css`
    label: logs-stats-row--active;
    color: ${theme.colors.primary.text};
    position: relative;
  `,
  logsStatsRowLabel: css`
    label: logs-stats-row__label;
    display: flex;
    margin-bottom: 1px;
  `,
  logsStatsRowValue: css`
    label: logs-stats-row__value;
    flex: 1;
    text-overflow: ellipsis;
    overflow: hidden;
  `,
  logsStatsRowCount: css`
    label: logs-stats-row__count;
    text-align: right;
    margin-left: 0.5em;
  `,
  logsStatsRowPercent: css`
    label: logs-stats-row__percent;
    text-align: right;
    margin-left: 0.5em;
    width: 3em;
  `,
  logsStatsRowBar: css`
    label: logs-stats-row__bar;
    height: 4px;
    overflow: hidden;
    background: ${theme.colors.text.disabled};
  `,
  logsStatsRowInnerBar: css`
    label: logs-stats-row__innerbar;
    height: 4px;
    overflow: hidden;
    background: ${theme.colors.primary.main};
  `
});
const LogLabelStatsRow = ({ active, count, proportion, value }) => {
  const style = useStyles2(getStyles);
  const percent = `${Math.round(proportion * 100)}%`;
  const barStyle = { width: percent };
  const className = active ? cx([style.logsStatsRow, style.logsStatsRowActive]) : cx([style.logsStatsRow]);
  return /* @__PURE__ */ React__default.createElement("div", { className }, /* @__PURE__ */ React__default.createElement("div", { className: cx([style.logsStatsRowLabel]) }, /* @__PURE__ */ React__default.createElement("div", { className: cx([style.logsStatsRowValue]), title: value }, value), /* @__PURE__ */ React__default.createElement("div", { className: cx([style.logsStatsRowCount]) }, count), /* @__PURE__ */ React__default.createElement("div", { className: cx([style.logsStatsRowPercent]) }, percent)), /* @__PURE__ */ React__default.createElement("div", { className: cx([style.logsStatsRowBar]) }, /* @__PURE__ */ React__default.createElement("div", { className: cx([style.logsStatsRowInnerBar]), style: barStyle })));
};
LogLabelStatsRow.displayName = "LogLabelStatsRow";

export { LogLabelStatsRow };
