// @grafana/faro-core/dist/spec/core/src/api/logs/initialize.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var initialize_1 = require("../../initialize");
var testUtils_1 = require("../../testUtils");
var utils_1 = require("../../utils");
describe('api.logs', function () {
    function createAPI(_a) {
        var _b = _a === void 0 ? { dedupe: true } : _a, dedupe = _b.dedupe;
        var transport = new testUtils_1.MockTransport();
        var config = (0, testUtils_1.mockConfig)({
            dedupe: dedupe,
            transports: [transport],
        });
        var api = (0, initialize_1.initializeFaro)(config).api;
        return [api, transport];
    }
    describe('pushLog', function () {
        var api;
        var transport;
        beforeEach(function () {
            var _a;
            _a = createAPI(), api = _a[0], transport = _a[1];
        });
        describe('Filtering', function () {
            it('filters the same event', function () {
                api.pushLog(['test']);
                expect(transport.items).toHaveLength(1);
                api.pushLog(['test']);
                expect(transport.items).toHaveLength(1);
            });
            it("doesn't filter events with partially same message", function () {
                api.pushLog(['test', 'another test']);
                expect(transport.items).toHaveLength(1);
                api.pushLog(['test']);
                expect(transport.items).toHaveLength(2);
            });
            it("doesn't filter events with same message and different level", function () {
                api.pushLog(['test']);
                expect(transport.items).toHaveLength(1);
                api.pushLog(['test'], {
                    level: utils_1.LogLevel.INFO,
                });
                expect(transport.items).toHaveLength(2);
            });
            it("doesn't filter events with same message and same level but different context", function () {
                api.pushLog(['test']);
                expect(transport.items).toHaveLength(1);
                api.pushLog(['test'], {
                    context: {
                        a: '1',
                    },
                });
                expect(transport.items).toHaveLength(2);
            });
            it("filters an event and doesn't filter the next different one", function () {
                api.pushLog(['test']);
                expect(transport.items).toHaveLength(1);
                api.pushLog(['test']);
                expect(transport.items).toHaveLength(1);
                api.pushLog(['test'], {
                    level: utils_1.LogLevel.ERROR,
                });
                expect(transport.items).toHaveLength(2);
            });
            it("doesn't filter when dedupe is false", function () {
                var _a;
                _a = createAPI({ dedupe: false }), api = _a[0], transport = _a[1];
                api.pushLog(['test']);
                expect(transport.items).toHaveLength(1);
                api.pushLog(['test']);
                expect(transport.items).toHaveLength(2);
            });
            it("doesn't filter when skipDedupe is true", function () {
                api.pushLog(['test']);
                expect(transport.items).toHaveLength(1);
                api.pushLog(['test'], { skipDedupe: true });
                expect(transport.items).toHaveLength(2);
            });
        });
    });
});
